crates/chain/src/keychain/txout_index.rs

```diff
@@ -367,60 +525,70 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     ///
     /// This will panic if the given `keychain`'s descriptor does not exist.
     pub fn unbounded_spk_iter(&self, keychain: &K) -> SpkIterator<Descriptor<DescriptorPublicKey>> {
-        SpkIterator::new(
-            self.keychains
-                .get(keychain)
-                .expect("keychain does not exist")
-                .clone(),
-        )
+        let descriptor = self
+            .descriptor_of_keychain(keychain)
+            .expect("Keychain must exist")
+            .1
+            .clone();
+        SpkIterator::new(descriptor)
     }
 
     /// Get unbounded spk iterators for all keychains.
     pub fn all_unbounded_spk_iters(
         &self,
     ) -> BTreeMap<K, SpkIterator<Descriptor<DescriptorPublicKey>>> {
-        self.keychains
+        self.keychains_to_descriptors
             .iter()
-            .map(|(k, descriptor)| (k.clone(), SpkIterator::new(descriptor.clone())))
+            .map(|(k, (_, descriptor))| (k.clone(), SpkIterator::new(descriptor.clone())))
             .collect()
     }
 
     /// Iterate over revealed spks of all keychains.
     pub fn revealed_spks(&self) -> impl DoubleEndedIterator<Item = (K, u32, &Script)> + Clone {
-        self.keychains.keys().flat_map(|keychain| {
+        self.keychains_to_descriptors.keys().flat_map(|keychain| {
             self.revealed_keychain_spks(keychain)
                 .map(|(i, spk)| (keychain.clone(), i, spk))
         })
     }
 
     /// Iterate over revealed spks of the given `keychain`.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if the given `keychain`'s descriptor does not exist.
     pub fn revealed_keychain_spks(
         &self,
         keychain: &K,
     ) -> impl DoubleEndedIterator<Item = (u32, &Script)> + Clone {
-        let next_i = self.last_revealed.get(keychain).map_or(0, |&i| i + 1);
+        let desc_id = self.descriptor_of_keychain(keychain).expect("Must exist").0;
+        let next_i = self.last_revealed.get(&desc_id).map_or(0, |&i| i + 1);
         self.inner
             .all_spks()
-            .range((keychain.clone(), u32::MIN)..(keychain.clone(), next_i))
+            .range((desc_id, u32::MIN)..(desc_id, next_i))
             .map(|((_, i), spk)| (*i, spk.as_script()))
     }
 
     /// Iterate over revealed, but unused, spks of all keychains.
     pub fn unused_spks(&self) -> impl DoubleEndedIterator<Item = (K, u32, &Script)> + Clone {
-        self.keychains.keys().flat_map(|keychain| {
+        self.keychains_to_descriptors.keys().flat_map(|keychain| {
             self.unused_keychain_spks(keychain)
                 .map(|(i, spk)| (keychain.clone(), i, spk))
         })
     }
 
     /// Iterate over revealed, but unused, spks of the given `keychain`.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if the given `keychain`'s descriptor does not exist.
```

Comment:
Thanks for adding this comment! Could you add a similar comment to `lookahead_to_target` as well?

I'm also thinking whether panicing when the keychain does not exist is the best move. @LLFourn what are your thoughts? I think in anycase, this can be tackled in a separate PR because this was already the old behavior.

Reply:
I don't think it's a good idea to return an empty iterator here. If the user passed in a non-existent keychain they probably want to know about it, and panicking is a really explicit way (probably too explicit!) of doing so, while returning an empty iterator is not explicit at all (in this particular case, not having any unused spks or passing in a wrong keychain yields the same exact result!)

Reply:
In this specific case they are querying for unused spks of a keychain. If the keychain has not revealed any spks yet it should be empty. I think this is the right answer regardless of whether it has had a descriptor associated with the keychain yet. I don't think there is a useful concept of a "wrong keychain" -- just a keychain that doesn't have descriptor associated with it yet.

Consider my current in-progress wallet architecture for frostsnap:

1. When we generate a new key we just insert it into the db (but not the `KeychainTxOutIndex`).
2. In general when the wallet gets queried forwards to the `IndexedTxGraph` / `LocalChain` and transforms it just like bdk's `Wallet`.
3. When the application needs a new address, we check whether we have already got the descriptor in the index, [otherwise we insert it](https://github.com/frostsnap/frostsnap/pull/113/files#diff-57f8639fc9052a85c8cff9c3d5ebe6f20e9a87105d759bb8996fba159acc3e21R218) by seeing if it exists in the database.

I don't think I should be forced to do this check-and-insert before every query to the `KeychainTxOutIndex`. If somehow my app queries `unused_keychain_spks` to the backend before calling an spk deriving method, It'd be much better for me if it just returned that there are no unused spks because that's the correct answer.

The new spk derivation methods do need the descriptor of course and we have to decide whether to panic or to error there. The answer depends on whether in all cases, making a new spk query without having inserted the descriptor represents a programmer error. It's hard to make this judgement for all contexts so it's probably better to indicate in the return type the possibility of the keychain's descriptor not existing. I guess an `Option` would be best for this.